fairscale/nn/moe/top2gate.py

```diff
@@ -0,0 +1,111 @@
+    mask1 = F.one_hot(indices1_gs, num_classes=num_experts)
+
+    # Create a mask for 2nd's expert per token using Gumbel-max trick
+    # https://timvieira.github.io/blog/post/2014/07/31/gumbel-max-trick/
+    logits_w_noise = logits + gumble.rsample(logits.shape)
+    # Replace top-expert with min value
+    mins = torch.full_like(logits, min_logit)
+    logits_except1 = torch.where(mask1.bool(), mins, logits_w_noise)
+    indices2_gs = torch.argmax(logits_except1, dim=2)
+    mask2 = F.one_hot(indices2_gs, num_classes=num_experts)
+
+    # Compute locations in capacity buffer
+    locations1 = torch.cumsum(mask1, dim=1) - 1
+    locations2 = torch.cumsum(mask2, dim=1) - 1
+    # Update 2nd's location by accounting for locations of 1st
+    locations2 += torch.sum(mask1, dim=1, keepdim=True)
+
+    # Compute l_aux
+    me = torch.mean(gates, dim=1)
+    ce = torch.mean(mask1.float(), dim=1)
+    l_aux = torch.mean(me * ce)
+
+    # Remove locations outside capacity from mask
+    mask1 *= torch.lt(locations1, capacity)
+    mask2 *= torch.lt(locations2, capacity)
+
+    # Store the capacity location for each token
+    locations1_gs = torch.einsum("gse,gse->gs", locations1, mask1)
+    locations2_gs = torch.einsum("gse,gse->gs", locations2, mask2)
+
+    # Normalize gate probabilities
+    mask1_float = mask1.float()
+    mask2_float = mask2.float()
+    gates1_gs = torch.einsum("gse,gse->gs", gates, mask1_float)
+    gates2_gs = torch.einsum("gse,gse->gs", gates, mask2_float)
+    denom_gs = gates1_gs + gates2_gs
+    # Avoid divide-by-zero
+    denom_gs = torch.where(denom_gs > 0, denom_gs, torch.ones_like(denom_gs))
+    gates1_gs /= denom_gs
+    gates2_gs /= denom_gs
+
+    # Calculate combine_weights and dispatch_mask
+    gates1 = torch.einsum("gs,gse->gse", gates1_gs, mask1_float)
+    gates2 = torch.einsum("gs,gse->gse", gates2_gs, mask2_float)
+    locations1_gsc = F.one_hot(locations1_gs, num_classes=capacity)
+    locations2_gsc = F.one_hot(locations2_gs, num_classes=capacity)
+    combine1_gsec = torch.einsum("gse,gsc->gsec", gates1, locations1_gsc)
+    combine2_gsec = torch.einsum("gse,gsc->gsec", gates2, locations2_gsc)
+    combine_weights = combine1_gsec + combine2_gsec
+    dispatch_mask = combine_weights.bool()
+
+    return l_aux, combine_weights, dispatch_mask
+
+
+class Top2Gate(torch.nn.Module):
+    """Gate module which implements Top2Gating as described in Gshard_.
+    ::
+
+        gate = Top2Gate(model_dim, num_experts)
+        l_aux, combine_weights, dispatch_mask = gate(input)
+
+    .. Gshard_: https://arxiv.org/pdf/2006.16668.pdf
+
+    Args:
+        model_dim (int):
+            size of model embedding dimension
+        num_experts (ints):
+            number of experts in model
+    """
+
+    wg: torch.nn.Linear
+
+    def __init__(self, model_dim: int, num_experts: int,) -> None:
+        super().__init__()
+        self.wg = torch.nn.Linear(num_experts, model_dim, bias=False)
```

Comment:
nit: extract this to separate module? don't need to make it configurable for now though

Reply:
so:

def __init__(self, wg: torch.nn.Linear) -> None: